### PR TITLE
Move Program.Main into WebApplication.Start.

### PR DESCRIPTION
**Pull request by Tratcher**
#430 @davidfowl @DamianEdwards @muratg 
This is a really strange class and it's going to confuse anyone that used Katana. It's especially strange that Start never returns. Can we rename Start to Main or Run?

### CODE REVIEW

**Comment by muratg** [comment]
Perhaps we should keep the Program class, and add the Start methods there?

**Comment by Tratcher** [comment]
`public static void Main(string[] args) => Program.Start<Startup>(args);`? Hmm, I think that looses some context because you don't see who's Program your calling. Nor does it address the name Start.

**Comment by davidfowl** [comment]
Call it run instead of start

**Comment by davidfowl** [comment]
Also optionally take in the startup class

**Comment by Tratcher** [comment]
An instance of the startup class rather than just the type? We don't actually support that anywhere in Hosting right now and it won't be trivial to add. We can work on that later.

**Comment by davidfowl** [comment]
@Tratcher no, WebApplication.Run<TStartup>()

**Comment by davidfowl** [comment]
Oh wait you have that :small_blue_diamond:

**Comment by davidfowl** [comment]
Rebase and :shipit:

**Comment by davidfowl** [comment]
Hmmm hold on

File: src/Microsoft.AspNet.Hosting/WebApplication.cs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        }
+
+        public static void Start(Type startupType, string[] args)
+        {
+            // Allow the location of the json file to be specified via a --config command line arg
+            var tempBuilder = new ConfigurationBuilder().AddCommandLine(args);
+            var tempConfig = tempBuilder.Build();
+            var configFilePath = tempConfig[ConfigFileKey] ?? HostingJsonFile;
+
+            var appBasePath = PlatformServices.Default.Application.ApplicationBasePath;
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Hosting/WebApplication.cs]: We need to tweak this a bit. Use this logic:

https://github.com/aspnet/Configuration/blob/dev/src/Microsoft.Extensions.Configuration.FileExtensions/FileConfigurationExtensions.cs#L52-L57

To resolve the application base path

File: src/Microsoft.AspNet.Hosting/Program.cs [hunk 2]
```diff
@@ -1,65 +1,13 @@
...
-                Console.CancelKeyPress += (sender, eventArgs) =>
-                {
-                    appLifetime.StopApplication();
-                    // Don't terminate the process immediately, wait for the Main thread to exit gracefully.
-                    eventArgs.Cancel = true;
-                };
-
-                appLifetime.ApplicationStopping.WaitHandle.WaitOne();
-            }
+            WebApplication.Start(args);
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Hosting/Program.cs]: Rename Start to Run

File: src/Microsoft.AspNet.Hosting/WebApplication.cs (hunk 1 above)
Comment (Tratcher) [inline src/Microsoft.AspNet.Hosting/WebApplication.cs]: Why isn't that done inside of PlatformServices.Default.Application.ApplicationBasePath?

**Comment by davidfowl** [comment]
:shipit:

**Pull request merged**